x-pack/plugins/siem/public/common/store/index.ts

```diff
@@ -8,18 +8,37 @@ export * from './model';
 export * from './reducer';
 export * from './selectors';
 
+import { Middleware, Dispatch } from 'redux';
 import { createStore, getStore } from './store';
-import { SubstateMiddlewareFactory } from './types';
+import { ImmutableMiddleware } from './types';
+import { AppAction } from './actions';
+import { State } from './reducer';
+import { Immutable } from '../../../common/endpoint/types';
 
 export { createStore, getStore };
 
-export const substateMiddlewareFactory: SubstateMiddlewareFactory = (selector, middleware) => {
+/**
+ * Takes a selector and an `ImmutableMiddleware`. The
+ * middleware's version of `getState` will receive
+ * the result of the selector instead of the global state.
+ *
+ * This allows middleware to have knowledge of only a subsection of state.
+ *
+ * `selector` returns an `Immutable` version of the substate.
+ * `middleware` must be an `ImmutableMiddleware`.
+ *
+ * Returns a regular middleware, meant to be used with `applyMiddleware`.
+ */
+export const substateMiddlewareFactory = <Substate>(
+  selector: (state: State) => Substate | Immutable<Substate>,
+  middleware: ImmutableMiddleware<Substate, AppAction>
+): Middleware<{}, State, Dispatch<AppAction | Immutable<AppAction>>> => {
   return (api) => {
     const substateAPI = {
       ...api,
       // Return just the substate instead of global state.
-      getState() {
-        return selector(api.getState());
+      getState(): Immutable<Substate> {
+        return selector(api.getState()) as Immutable<Substate>;
```

Comment:
#### re: the `selector` can return a mutable version
You mean by "cloning" the value (if its an object), correct? but even then, I would assume we would need some sort of a `Mutable<>` util type to change everything back to mutable when typing up such a return value - no?

Reply:
Actually my code comment is not exactly correct anymore. The immutable code was confusing me. Here's the thing:

For convenience, we let `ImmutableReducer` return a mutable or immutable version of state. That means that the type return value for `ImmutableReducer<S>` is `S | Immutable<S>`.
If a reducer clones state, and then mutates it, and then returns that, we allow it. It's safe. If we said that the return value of `ImmutableReducer` was `Immutable<S>`, the types could be a lot simpler to understand, and they would work better, but you'd have to cast the return value to `Immutable` in many reducers.

e.g.
```typescript
const r: ImmutableReducer<{ a : string }, AnyAction> = (state, action) => {
  // This matches `{ a: string }` structurally, but its not explicitly `Immutable` so the return type
  // could be inferred as `{ a : string }` but not `Immutable<{ a: string }>`
  return {
    ...state,
    a: "hi"
  }
}
```

This means that the selector will receive a mutable version of state (above this point, state is mutable.) This selector casts it to `Immutable`, so from this point on, it will be `Immutable`. This is sorta like the `Immutable` diode.

This immutable stuff might've been a mistake.

If you use a type like `ImmutableReducer<T>` with redux types, it'll infer it to be `Reducer<T | Immutable<T>>`. That's because a redux reducer (proper one) takes and returns the same state type:

```
export type Reducer<S = any, A extends Action = AnyAction> = (
  state: S | undefined,
  action: A
) => S
```

And `Immutable<T> | T` satisfies `S` in both the parameter position and the return position. Imagine if you had `ImmutableReducer<T | Immutable<T>>`. Its return type would be: `Immutable<T> | T`. Because `Immutable<Immutable<T>` is just `Immutable<T>`.

For this reason, you'll end up w/ the top level state being inferred as stuff like `Immutable<ManagementState> | ManagementState`.

For example:
![image](https://user-images.githubusercontent.com/35559/83536902-dfe21300-a4c1-11ea-8ebf-d7b5bf3e175c.png)

With this PR, I define top level state for all endpoint subplugins as the base type (not the immutable one.)

Reply:
Thanks for the long explanation. I'm dizzy.